Bucket-processor : do not retry the whole listing range when their is an error on individual objects

francoisferrand opened the pull request:
Misc tweaks to lifecycle, which should avoid duplicated processing.

- Properly handle async calls to `_applyTransitionRule`
- Do not retry whole batch when a single entry fails

Issue: BB-437

francoisferrand committed:
Properly handle async calls to _applyTransitionRule

This esnures we don't process more entries than the expected concurrency
at the same time.

Issue: BB-437

francoisferrand committed:
Do not retry whole batch when a single entry fails

We don't want to retry the _whole_ list if only a single entry fails, so
we possibly retry each entry individually, and ignore errors. Ignoring
error is not too bad anyway, the entry will be picked up again on next
lifecycle run.

Issue: BB-437

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-437 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.28`

* `8.7.0`


Please check the `Fix Version/s` of BB-437, or the target
branch of this pull request.

nicolas2bert commented:
Thanks, could it target development/7.10 branch?

nicolas2bert commented on extensions/lifecycle/tasks/LifecycleTask.js:
```diff
@@ -894,18 +894,32 @@ class LifecycleTask extends BackbeatTask {
-                    }
-                    return this._compareObject(bucketData, obj, applicableRules, log,
-                        next);
+            // We don't want to retry the _whole_ list if only a single entry fails,
```
The same logic has to be applied to `LifecycleTaskV2._compareRulesToList()`.

nicolas2bert commented on extensions/lifecycle/tasks/LifecycleTask.js:
```diff
@@ -894,18 +894,32 @@ class LifecycleTask extends BackbeatTask {
-                    return this._compareObject(bucketData, obj, applicableRules, log,
-                        next);
+            // We don't want to retry the _whole_ list if only a single entry fails,
+            // so we possibly retry each individual entry here, and ignore errors.
```
Should we even retry? Retry has exponential backoff and could take seconds to complete (to be checked).

francoisferrand replied:
I think we should retry, at least until lifecycle optimization is used everywhere and we can run lifecycle more often : otherwise a minor retriable error could delay lifecycle for 12h.

That said, indeed retry has delay which we need to consider, esp. w/ respect to being a kafka consumer.

francoisferrand replied:
actually, looking at the code of `_applyTransitionRule()`, we already skip the errors from `s3:_getObjectMD`... though some errors can still go through, e.g. when  calling in `s3:listObjectVersions` ( `_checkAndApplyEODMRule()`) or `s3:headObject` in non-versionned bucket...

nicolas2bert commented on extensions/lifecycle/tasks/LifecycleTask.js:
```diff
@@ -1469,16 +1483,15 @@ class LifecycleTask extends BackbeatTask {
             }
             if (rules.Transition) {
-                this._applyTransitionRule({
+                return this._applyTransitionRule({
```
While I believe it's a beneficial concept to implement asynchronous logic for better flow control, there's a possibility that it could introduce a delay in the execution of the bucket processor task, as it would need to wait for the callback to complete, potentially affecting the Kafka consumer (as discussed earlier).

nicolas2bert replied:
This could clarify why the processing of Lifecycle bucket tasks doesn't seem very asynchronous – we weren't waiting for the callback to be invoked. This might be due to concerns about its potential impact on the Kafka consumer.

francoisferrand replied:
individual objects are processed in parallel (hard-coded to 10 at a time), so it should be fine : and if this introduces a delay, we will need to _manage_ it, not just run uncontrolled number of operations in parallel.
Another hint is that the callback is used in lifecycle v2 task.

rachedbenmustapha commented on extensions/lifecycle/conductor/LifecycleConductor.js:
```diff
@@ -747,12 +747,18 @@ class LifecycleConductor {
                 this.lcConfig.objectProcessor.groupId, 0, next),
 
+            // check that object tasks topic consumer lag for transitions is 0
+            objectTasksTransitionCheck:
```
There might be several topics to check for transition, as each cold location has its own.

francoisferrand replied:
After discussion, it seems this may not be appropriate after all:
* the original goal of checking lag is to ensure we do not start a new listing until the previous one has been terminated, i.e. that each bucket identified by the conductor has been listed → so we should only check that the listing is complete, i.e. no lag on bucket topic
* checking too many topics may make unexpectedly skip the iteration : which is fine when the cron job runs every few minute, and kind of acceptable when the system is still busy processing the last batch... though it may become unacceptable if "other action" (restore expiration, restoration, end of archive...) were to delay the listing, possibly by 
* processing expiration is fast and transitions update the metadata, but it may be better to delay next listing anyway : to avoid filling the topic with objects which are already in it
* yet we certainly don't want the next listing to be delayed by operations not directly related to the listing : restore request, restore expiration, transition status update, ...

**NOTE**: If the cronjob is not running very often, skipping a run may have a significant impact. We should probably generate an alert if the scan lasts longer than the cronjob period, and/or if the scan period is more than 24h → https://scality.atlassian.net/browse/ZENKO-4665

francoisferrand replied:
Topics/consumers groups to check identified, but there may be more work needed: moved to https://scality.atlassian.net/browse/BB-449

francoisferrand committed:
gha: Bump setup-node to v3

Issue: BB-437

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-437 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.29`

* `8.7.0`


Please check the `Fix Version/s` of BB-437, or the target
branch of this pull request.

francoisferrand committed:
Bucket-processor: limit the amount of retries

Limit the total amount of retries when processing a range, to keep the
processing time of the kafka message under check (--> avoid rebalance),
yet keep some resiliance to temporary errors.

It does not really matter that we don't retry a lot, since lfiecycle
will be retried later.

Aslo, propagate the error from _applyTransitionRule() so it may get
retried.

Issue: BB-437

francoisferrand committed:
Better handle errors on individual objects in lifecycle v2

And let `_compareRulesToList()` retry just a little bit, to limit impact/delay
while ensuring message gets processed timely.

Issue: BB-437

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-437 contains:


* `8.6.28`

* `8.7.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.29`

* `8.7.0`


Please check the `Fix Version/s` of BB-437, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

benzekrimaha commented on extensions/lifecycle/tasks/LifecycleTask.js:
```diff
@@ -35,6 +35,20 @@ const MAX_KEYS = process.env.CI === 'true' ? 3 : 1000;
+// seconds.
+const MAX_RETRIES = 4;
+
+// Maximum number of retries when processing entries in the rang
```
```suggestion
// Maximum number of retries when processing entries in the range
```

francoisferrand commented:
/approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-437` into
`development/8.6`
* [w/8.7/bugfix/BB-437](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/bugfix/BB-437) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



Please check the status of the associated issue BB-437.

Goodbye francoisferrand.
